1237 let the user select a prover

eugenk opened the pull request:
This shall fix #1237. This branch allows the user to select _multiple_ available provers for proving. For each prover, a proof attempt is made. Those proof attempts are processed sequentially.
#### Possibly important notes
- The `prove/` route has been replaced by `proofs/` to use the REST convention
- The new ProofsController is completely rewritten and has nothing in common with the previous ProveController
- I tried to make many small commits and to not revert previous changes in the same branch. It should be easier to review this pull request commit by commit.

This branch is based on the branch of #1262 (**introduce_hets_options**) instead of **staging**.  It might be easier to review that one first.

eugenk committed:
Allow to prove a single theorem.

eugenk committed:
Add model: Prover.

eugenk committed:
Associate Prover with OntologyVersion.

eugenk committed:
Retrieve provers for an ontology.

eugenk committed:
Bump minimal hets version

This version introduces JSON output for the provers list.

eugenk committed:
Always stub the provers call along with the dg call.

eugenk committed:
Fix typo in specroutes doc.

eugenk committed:
Fix routing: Put axioms, theorems before sentences.

eugenk committed:
Use axioms for used_axioms.

eugenk committed:
Fix ontology iri passed to hets api.

eugenk committed:
Add ProversOptions class.

eugenk committed:
Add method: provers_via_api.

eugenk committed:
Add spec for proving a single theorem.

eugenk committed:
Add Provers Evaluator

The evaluator uses JSON.parse because it always operates on short
strings. No need for a SAX parser here.

eugenk committed:
Add ProversCaller.

eugenk committed:
Add spec for provers retrieval when parsing.

eugenk committed:
Use Prover model in ProofAttempt instead its name.

eugenk committed:
Adjust spec to use prover model.

eugenk committed:
Adjust factory to use Prover model.

eugenk committed:
Add prover to ProveOptions.

eugenk committed:
Pass ProveOptions to [async_]prove methods.

eugenk committed:
Add CollectiveProofAttempt[Worker]

This classes allow to (asynchronously) make many proof attempts
sequentially. In a later phase, this will be reconstructed to work in
parallel.

eugenk committed:
Rename/substantivize ProveController and its route.

eugenk committed:
Add route, action and views for proofs#new.

eugenk committed:
Add ProversOptions spec.

eugenk committed:
Rewrite ProofsController spec.

eugenk committed:
Add Proof FakeRecord.

eugenk committed:
Always show prove button.

eugenk committed:
Rewrite ProofsController to use Proof FakeRecord.

eugenk committed:
Add CollectiveProofAttempt spec.

eugenk committed:
Add Proof spec.

eugenk committed:
Add CollectiveProofAttemptWorker spec.

tillmo commented:
when trying to prove the goal in `strict_partial_order` with SPASS, I get `NoMethodError: undefined method "prove" for #<Theorem:0x00000006b682e8>`

eugenk commented:
That's odd. Have you restarted the rails server after switching the branch to 1237-let_the_user_select_a_prover?

tillmo commented:
yes. But I forgot to restart sidekiq - that solved the problem. Is there a simple way to do all these restarts with one command?

eugenk commented:
Yes, you can stop invoker and start it again.
During development, we start everything with
```
invoker start invoker.ini
```
from the ontohub directory. Whenever we need to restart the web server and sidekiq, we simply hit <kbd>Crtl+C</kbd> and start invoker again.

tillmo commented:
when I prove `asymmetric` in `strict_partial_order` with `eprover`, I get "Time taken -1 seconds". This is extremely fast...

tillmo commented:
when I select several provers at once (i.e. for a CASL goal: leo, eprover, darwin, SPASS, Vampire, leoII,...), I get in the job queue:
```
 less than a minute from now 	0 	hets 	CollectiveProofAttemptWorker 	
"OntologyVersion", 28, [7, 8, 9, 10, 11, 12, 13, 18]   JSON::Stream::ParserError: Expected null keyword: char 1
```
If I select just one prover (SPASS), the theorem gets proven immediately.

eugenk commented:
Yes, this seems to be a problem in hets, because it returns:
```json
 "usedTime": {
  "seconds": -1,
  "components": { "hours": -1, "mins": 59, "secs": 59 }
  }
```
I can change ontohub to save 0 seconds if hets gives us a negative value.

tillmo commented:
aha, the problem is not with selecting multiple provers, but with selecting leo. If leo is among the provers, the error occurs.

eugenk commented:
That's too bad, I don't have leo on my machine. Could you please post the outputs of
```
curl -X GET http://localhost:8000/version
curl -X POST -d '{"format":"json","include":"true","prover":"leo","node":"strict_partial_order"}' -H 'Content-Type: application/json' http://localhost:8000/prove/http%3A%2F%2Flocalhost%3A3000%2Fref%2F1%2Fdefault%2Fpartial_order%2F%2Fstrict_partial_order/auto
```
This should attempt to prove `strict_partial_order` with leo.

tillmo commented:
selecting QucikCheck leads to `EOFError: end of file reached`.
isabelle-process and isabelle-jedit lead to `JSON::Stream::ParserError: Expected null keyword: char 1`

tillmo commented:
problem with leo solved. I has installed leo, but not ocaml. There should be a proper Ubuntu package for leo...

tillmo commented:
the interactive Isabelle provers (with jedit, emacs etc.) should be removed from the list, because it is an own project to integrate them into Ontohub (see #136). I think Hets should provide some flag whether the prover is automatic or interactive.

tillmo commented:
after selecting Isabellesledgehammer from the list of provers, I get a new prover in the list: Isabelle (sledgehammer), while Isabellesledgehammer remains in the list.

tillmo commented:
When I prove the goals in http://develop.ontohub.org/sandbox/Family2.het and select Pellet and Fact, I get a "THM", but the link links to a SPASS proof of a CASL theorem in strict_partial_order. Due to #1277, I could only reproduce this locally.

tillmo commented:
in order to provide more details: here is the list of theorems that I have:
```
[2] pry(main)> Theorem.all
  Theorem Load (0.8ms)  SELECT "sentences".* FROM "sentences" WHERE "sentences"."imported" = 'f' AND "sentences"."type" IN ('Theorem')
=> [#<Theorem id: 1239, ontology_id: 5, name: "asymmetric", text: "forall x, y : elem . x < y => not y < x", range: "http://localhost:3000/ref/3/default/partial_order:1...", comments_count: 0, created_at: "2015-03-21 19:26:55", updated_at: "2015-03-21 20:52:30", display_text: "forall x, y : elem . x < y => not y < x", imported: false, type: "Theorem", proof_status_id: "THM", locid: "/default/partial_order//strict_partial_order//asymm...">,
 #<Theorem id: 1241, ontology_id: 6, name: "reflexive", text: "forall x : elem . x <= x", range: "http://localhost:3000/ref/3/default/partial_order:5...", comments_count: 0, created_at: "2015-03-21 19:26:55", updated_at: "2015-03-21 19:26:55", display_text: "forall x : elem . x <= x", imported: false, type: "Theorem", proof_status_id: "OPN", locid: "/default/partial_order//v__T//reflexive">,
 #<Theorem id: 1242, ontology_id: 6, name: "antisymmetric", text: "forall x, y : elem . x <= y /\\ y <= x => x = y", range: "http://localhost:3000/ref/3/default/partial_order:5...", comments_count: 0, created_at: "2015-03-21 19:26:55", updated_at: "2015-03-21 19:26:55", display_text: "forall x, y : elem . x <= y /\\ y <= x => x = y", imported: false, type: "Theorem", proof_status_id: "OPN", locid: "/default/partial_order//v__T//antisymmetric">,
 #<Theorem id: 1243, ontology_id: 6, name: "transitive", text: "forall x, y, z : elem . x <= y /\\ y <= z => x <= z", range: "http://localhost:3000/ref/3/default/partial_order:5...", comments_count: 0, created_at: "2015-03-21 19:26:55", updated_at: "2015-03-21 19:26:55", display_text: "forall x, y, z : elem . x <= y /\\ y <= z => x <= z", imported: false, type: "Theorem", proof_status_id: "OPN", locid: "/default/partial_order//v__T//transitive">,
 #<Theorem id: 1254, ontology_id: 21, name: "reflexive", text: "forall x : elem . x <= x", range: "http://localhost:3000/ref/1/default/partial_order2:...", comments_count: 0, created_at: "2015-03-21 19:28:29", updated_at: "2015-03-21 19:28:29", display_text: "forall x : elem . x <= x", imported: false, type: "Theorem", proof_status_id: "OPN", locid: "/default/partial_order2//v__T//reflexive">,
 #<Theorem id: 1255, ontology_id: 21, name: "antisymmetric", text: "forall x, y : elem . x <= y /\\ y <= x => x = y", range: "http://localhost:3000/ref/1/default/partial_order2:...", comments_count: 0, created_at: "2015-03-21 19:28:29", updated_at: "2015-03-21 19:28:29", display_text: "forall x, y : elem . x <= y /\\ y <= x => x = y", imported: false, type: "Theorem", proof_status_id: "OPN", locid: "/default/partial_order2//v__T//antisymmetric">,
 #<Theorem id: 1256, ontology_id: 21, name: "transitive", text: "forall x, y, z : elem . x <= y /\\ y <= z => x <= z", range: "http://localhost:3000/ref/1/default/partial_order2:...", comments_count: 0, created_at: "2015-03-21 19:28:29", updated_at: "2015-03-21 19:28:29", display_text: "forall x, y, z : elem . x <= y /\\ y <= z => x <= z", imported: false, type: "Theorem", proof_status_id: "OPN", locid: "/default/partial_order2//v__T//transitive">,
 #<Theorem id: 1252, ontology_id: 20, name: "asymmetric", text: "forall x, y : elem . x < y => not y < x", range: "http://localhost:3000/ref/1/default/partial_order2:...", comments_count: 0, created_at: "2015-03-21 19:28:29", updated_at: "2015-03-21 19:31:59", display_text: "forall x, y : elem . x < y => not y < x", imported: false, type: "Theorem", proof_status_id: "THM", locid: "/default/partial_order2//strict_partial_order//asym...">,
 #<Theorem id: 1277, ontology_id: 23, name: "Ax1_19", text: "Class: Mother\n       SubClassOf: Parent", range: "http://localhost:3000/ref/1/default/Family:56.10-56...", comments_count: 0, created_at: "2015-03-21 20:38:35", updated_at: "2015-03-22 08:54:08", display_text: "Class: Mother SubClassOf: Parent", imported: false, type: "Theorem", proof_status_id: "UNK", locid: "/default/Family//FamilyBase//Ax1_19">,
 #<Theorem id: 1278, ontology_id: 23, name: "Ax2_20", text: "Individual: mary\n            Types: Grandmother", range: "http://localhost:3000/ref/1/default/Family:57.16-57...", comments_count: 0, created_at: "2015-03-21 20:38:35", updated_at: "2015-03-22 08:54:12", display_text: "Individual: mary Types: Grandmother", imported: false, type: "Theorem", proof_status_id: "THM", locid: "/default/Family//FamilyBase//Ax2_20">,
 #<Theorem id: 1279, ontology_id: 23, name: "Ax3_21", text: "Class: Man\n       EquivalentTo: Woman", range: "http://localhost:3000/ref/1/default/Family:58.10-59...", comments_count: 0, created_at: "2015-03-21 20:38:35", updated_at: "2015-03-22 08:54:12", display_text: "Class: Man EquivalentTo: Woman", imported: false, type: "Theorem", proof_status_id: "NOC", locid: "/default/Family//FamilyBase//Ax3_21">]
[3] pry(main)> 
```
Now, the link http://localhost:3000/repositories/default/ontologies/23/theorems/1278/proof_attempts/1 shows the correct ontology (FamilyBase), but "Proof attempt of asymmetric", where "asymmetric" is linked to http://localhost:3000/repositories/default/ontologies/23/theorems/1252, which actually does not exist, because it is a theorem id of strict_partial_order coupled with the ontology id of FamilyBase. It exists at http://localhost:3000/repositories/default/ontologies/20/theorems/1252/proof_attempts/1

tillmo commented:
Actually, for all new theorem that I prove, when displaying the proof details with the proof_attempts URL, I always get the proof details of theorem 1252 (asymmetric).

eugenk commented:
> selecting QucikCheck leads to `EOFError: end of file reached`.
isabelle-process and isabelle-jedit lead to `JSON::Stream::ParserError: Expected null keyword: char 1`

I added #1280 for it.

> the interactive Isabelle provers (with jedit, emacs etc.) should be removed from the list, because it is an own project to integrate them into Ontohub (see #136). I think Hets should provide some flag whether the prover is automatic or interactive.

I added #1281 for it.

> after selecting Isabellesledgehammer from the list of provers, I get a new prover in the list: Isabelle (sledgehammer), while Isabellesledgehammer remains in the list. Same with LeoII and Leo II.

I don't understand this. Maybe you could show it in the meeting today?

> Actually, for all new theorems that I prove, when displaying the proof details with the proof_attempts URL, I always get the proof details of theorem 1252 (asymmetric).

I'm working on it in the next branch (not pushed yet). There each ProofAttempt gets a locid and those are used for the links. Locally, this is working fine in the new branch.

eugenk commented:
This branch is now based on staging.

eugenk commented:
jenkins, test this please

tillmo commented on app/controllers/proofs_controller.rb:
```diff
@@ -0,0 +1,48 @@
+
+  def new
+    render template: 'proofs/new'
+  end
```
two new methods do not make sense

eugenk replied:
Fixed by 6a2fc9d

tillmo commented on app/models/ontology_version.rb:
```diff
@@ -17,6 +17,7 @@ class OntologyVersion < ActiveRecord::Base
   belongs_to :user
   belongs_to :ontology, :counter_cache => :versions_count
   belongs_to :commit
+  has_and_belongs_to_many :provers
```
please document the use of provers here. I assume that these are the provers that in principle can be used for proving with the ontology?

eugenk replied:
Commented in 940c9c3

tillmo commented on config/routes.rb:
```diff
@@ -243,15 +243,15 @@
+Will return a representation of the axiom. The axiom
 is determined according to the *locid.
-Currently this will return the list of all sentence of the ontology.
+Currently this will return the list of all axiom of the ontology.
```
all axiom*s*

eugenk replied:
Fixed by 0af4266

tillmo commented on lib/proof.rb:
```diff
@@ -0,0 +1,55 @@
+class Proof < FakeRecord
```
why isn't this a subclass of ActiveRecord? Shouldn't proofs be stored in the database?

eugenk replied:
Commented in 62aa918.

tillmo commented on lib/proof.rb:
```diff
@@ -0,0 +1,55 @@
+    @ontology = Ontology.find(opts[:ontology_id])
+    # HACK: remove the empty string from params
+    # Rails 4.2 introduces the html form option :include_hidden
+    @provers = opts[:proof][:provers].select(&:present?).map(&:to_i)
```
doesn't belong a hack massaging the params to the controller or a helper called from the controller?

eugenk replied:
I'm not sure whether it belongs to the "model" class or to the controller. My reason for putting it into the FakeRecord is that all the "real models" (ActiveRecord) do it by themselves and we don't need to handle it in the controller.

tillmo replied:
OK, understood.

tillmo commented on app/models/ontology_version/parsing.rb:
```diff
@@ -60,4 +61,21 @@ def parse_full
+    provers_io = Hets.provers_via_api(ontology, hets_options)
+    Hets::Provers::Evaluator.new(self, provers_io).import
+  end
 end
```
couldn't these methods be tested as well?

eugenk replied:
They are not explicitly tested, but implicitly by spec/models/ontology_version/provers_spec.rb

eugenk committed:
Remove duplicate new action.

eugenk committed:
Add comment on provers attribute.

eugenk committed:
Fix typos.

eugenk committed:
Add comment on Proof class.

eugenk commented:
Jenkins, test this please. (I will remove this comment as soon as the test job has been triggered)

tillmo commented:
:+1:
